compiler/optimizer/CopyPropagation.cpp

```diff
@@ -540,18 +544,32 @@ int32_t TR_CopyPropagation::perform()
                            prevDefNode,
                            prevDefNode->getDecimalPrecision());
 
+                        auto equivalentDefLookup = equivalentDefs.find(j);
                         if ((!thisSymRefIsPreferred &&
-                             (prevSymRefIsPreferred || (equivalentDefs[j-firstRealDefIndex] != j))))
+                             (prevSymRefIsPreferred || (equivalentDefLookup != equivalentDefs.end()))))
                            {
                            if (trace())
                               traceMsg(comp(), "000setting i %d to j %d\n", i, j);
-                           equivalentDefs[i-firstRealDefIndex] = equivalentDefs[j-firstRealDefIndex];
+                           if (equivalentDefLookup == equivalentDefs.end())
+                              {
+                              if (i != j)
+                                 equivalentDefs[i] = j;
+                              }
+                           else if (i == equivalentDefLookup->second)
+                              {
+                              equivalentDefs.erase(i);
+                              }
+                           else
+                              {
+                              equivalentDefs[i] = equivalentDefLookup->second;
+                              }
                            }
-                        else if (equivalentDefs[j-firstRealDefIndex] == j)
+                        else if (equivalentDefs.find(j) == equivalentDefs.end())
                            {
                            if (trace())
                                traceMsg(comp(), "111setting j %d to i %d\n", j, i);
-                           equivalentDefs[j-firstRealDefIndex] = i;
+                           if (i != j)
+                              equivalentDefs[j] = i;
                            }
```

Comment:
This check should be redundant in both cases, as `i` should not appear in any of the mappings so far. The first section does not make this assumption, adding this condition to manage it, whilst the second does.